src/pps/util.go

```diff
@@ -32,28 +32,26 @@ func init() {
 }
 
 // VisitInput visits each input recursively in ascending order (root last)
-func VisitInput(input *Input, f func(*Input)) {
+func VisitInput(input *Input, f func(*Input) error) error {
+	var source []*Input
 	switch {
 	case input == nil:
-		return // Spouts may have nil input
+		return nil // Spouts may have nil input
 	case input.Cross != nil:
-		for _, input := range input.Cross {
-			VisitInput(input, f)
-		}
+		source = input.Cross
 	case input.Join != nil:
-		for _, input := range input.Join {
-			VisitInput(input, f)
-		}
+		source = input.Join
 	case input.Group != nil:
-		for _, input := range input.Group {
-			VisitInput(input, f)
-		}
+		source = input.Group
 	case input.Union != nil:
-		for _, input := range input.Union {
-			VisitInput(input, f)
+		source = input.Union
+	}
+	for _, input := range source {
+		if err := VisitInput(input, f); err != nil {
+			return err
 		}
 	}
-	f(input)
+	return f(input)
```

Comment:
I think the common way we recognize `ErrBreak` is to catch it and return `nil`, in which case I think we'd need that logic to be in a wrapper function.